### DISCUSSION TITLE
Cannot populate Collection from custom Pivot Many-to-Many

### DISCUSSION THREAD

**Discussion by Lemas97** [mikro-orm/mikro-orm#3036] [Q&A]
I upgraded at v5.1.2 so I could use the custom pivot table as described [here](https://mikro-orm.io/docs/collections#custom-pivot-table-entity).

My entities are:
```
@Entity()
class Foo {
 ...

  @ManyToMany({ entity: () => Bar, pivotEntity: () => FooBarPivot }, 'foos')
    bars= new Collection<Bar>(this)
}

@Entity()
class Bar {
 ...

  @ManyToMany({ entity: () => Foo, pivotEntity: () => FooBarPivot }, 'bars')
    foos= new Collection<Foo>(this)
}
```
(also tried adding `{ owner: true }`)

and my pivot table class

```
@Entity()
export class FooBarPivot {
  @ManyToOne(() => Foo, { primary: true })
    foo: Foo

  @ManyToOne(() => Bar, { primary: true })
    bar: Bar

  @Property({ default: true })
    extraField: boolean
}
```

When I try 
`const populatedFoos = await em.findOneOrFail(Foo, { id }, { populate: ['bars'] })`

the bars are populated, but in the reverse side e.g.
`const populatedBars= await em.findOneOrFail(Bar, { id  }, { populate: ['foos'] })`

my `populatedBars.foos.getItems()` are always empty array.
I also tried `em.populate(populatedBars, ['foos'])`  and adding { owner: true }but same result.

So my data are populated from one side (foo.bars) but not from the other (bar.foos).
Thanks in advance.

**Versions**
| node | v14.17.1 |
| typescript | 4.4.4|
| mikro-orm | 5.1.2 |

**Comment by B4nan** [comment]
You either need the `owner: true` or drop the mapped by parameter from owning side. The entity definition from your example should actually produce a metadata error, as it would mean you define both sides as inverse. And you are mixing the two approaches for defining relations, which is most probably also causing issues.

```ts
@Entity()
class Foo {
 ...

  @ManyToMany({ entity: () => Bar, pivotEntity: () => FooBarPivot }) // <-- drop the mapped by attribute from owning side
    bars= new Collection<Bar>(this)
}

@Entity()
class Bar {
 ...

  @ManyToMany({ entity: () => Foo, pivotEntity: () => FooBarPivot, mappedBy: 'foos' }) // if you want to use the first parameter as options, rest of them will be ignored! dont mix those approaches
    foos= new Collection<Foo>(this)
}
```

If you need more help, please provide executable reproduction. We do have tests for populating inverse side m:n of custom pivot table entity relation, they are passing just fine. You can take a look at them for working example:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/composite-keys/custom-pivot-entity.sqlite.test.ts

**Reply by Lemas97** [reply]
Thank you, that worked.

I also have another one question about Custom Pivot table. Is there any way to use foo.bars.add(bar) with the extra field? e.g.

```typescript
foo.bars.add({ bar, false })
```

For now, if I want to change the default value of extraField, I do it by updating it this way:

 ```typescript
const example = await em.findOneOrFail(FooBarPivot, { foo: foo, bar: bar })
em.asing(example, { extraField: false })
await em.flush()
```

**Reply by B4nan** [reply]
Not via the m:n relation, you still need 1:m relation targetting the pivot entity if you want to read/write to the extra columns.

```ts
@Entity()
class Foo {

    @OneToMany({ entity: () => FooBarPivot, mappedBy: 'foo' })
    barPivots = new Collection<FooBarPivot>(this)

    @ManyToMany({ entity: () => Bar, pivotEntity: () => FooBarPivot }, 'foos')
    bars= new Collection<Bar>(this)

}
```

```ts
const item = em.create(FooBarPivot, { foo, bar, extraField: false });
foo.barPivots.add(item);
```

**Reply by hieupmzz** [reply]
@B4nan What if I want to filter on the pivotEntity?
For example: I want to get foo.bars with condition FooBarPivot.extraField is true

**Reply by B4nan** [reply]
You need to define the 1:m relations to it, that way you actually work with the pivot entity. When you work with the m:n you only use the pivot table to connect the tables - but they never exist on runtime.
